### ISSUE TITLE
SAVE/APPEND invalidates NetCDF file in Ferret v6.7

### ISSUE THREAD

**Issue by karlmsmith** [NOAA-PMEL/Ferret#1106]
**Reported by andrew wittenberg on 19 Apr 2011 00:15 UTC**
Appending to a NetCDF file invalidates the file!  ncdump version 4.0.1 says "Can't open HDF5 attribute", while ncdump version 3.6.1 says "Not a netCDF file".

```
        NOAA/PMEL TMAP
        FERRET v6.7  
        Linux rh5 (gfortran) 2.6.18-238.1.1.el5P - 04/04/11
        18-Apr-11 20:06     

yes? use coads_climatology
yes? let a = sst
yes? let b = 2*sst
yes? save/clob/file="a.nc" a
 LISTing to file a.nc
yes? sp ncdump -h a.nc   ! OK so far...
netcdf a {
dimensions:
        COADSX = 180 ;
        COADSY = 90 ;
        TIME = UNLIMITED ; // (12 currently)
variables:
        double COADSX(COADSX) ;
                COADSX:units = "degrees_east" ;
                COADSX:modulo = " " ;
                COADSX:point_spacing = "even" ;
                COADSX:axis = "X" ;
        double COADSY(COADSY) ;
                COADSY:units = "degrees_north" ;
                COADSY:point_spacing = "even" ;
                COADSY:axis = "Y" ;
        double TIME(TIME) ;
                TIME:units = "hour since 0000-01-01 00:00:00" ;
                TIME:time_origin = "01-JAN-0000 00:00:00" ;
                TIME:modulo = " " ;
                TIME:axis = "T" ;
        float A(TIME, COADSY, COADSX) ;
                A:missing_value = -1.e+34f ;
                A:_FillValue = -1.e+34f ;
                A:long_name = "SST" ;
                A:history = "From coads_climatology" ;

// global attributes:
                :history = "FERRET V6.7   18-Apr-11" ;
                :Conventions = "CF-1.0" ;
}
yes? save/append/file="a.nc" b
 LISTing to file a.nc
yes? sp ncdump -h a.nc   ! yikes
ncdump: a.nc: Can't open HDF5 attribute
```

This is a new bug, it wasn't in Ferret v6.7beta.

Note that this doesn't happen for all variables -- e.g. in the above example if we set A and B to 1, or to a pseudovariable like x[gx=1:10:1], we get no error.

Migrated-From: http://dunkel.pmel.noaa.gov/trac/ferret/ticket/1834

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 19 Apr 2011 01:13 UTC**
Ferret v67beta was linked with netCDF4.1.1 and an older hdf library.

I wonder if the netcdf/hdf versions need to be consistent between Ferret and ncdump in order for ncdump to be able to dump the header of an open file?  With the ncdump I have installed on our 32-bit linux machine, the above sequence does not result in an error, but Ferret v6.7 and ncdump4.1.1 also does not give the error.

For testing I'll send you a tar file with the ncdump and ncgen that came from the netCDF4.1.2 that ferret is linked with.

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 19 Apr 2011 19:11 UTC**
Indeed, the ncdump you sent me, built with NetCDF 4.1.2, works fine on the Ferret v6.7 output.

I'm shocked that the nominally minor increment from NetCDF 4.1.1->4.1.2 (third digit of the version number) would completely break Ferret compatibility with our existing tools (ncdump, ncview, ...), requiring a recompile of every tool built with the NetCDF 4.0.1 library.  Unidata should probably have given more thought to their version names!

I'm giving up testing of Ferret v6.7 for now, returning to v6.7beta until we can install NetCDF 4.1.2 at GFDL and recompile our tools...

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 19 Apr 2011 19:15 UTC**
Perhaps it's worth setting up a regression test for this -- e.g. doing some SAVE/APPEND's and then trying to "ncdump -h" them using previous versions of ncdump.  In the future, it'll be good to know when Ferret users will need to recompile their tools due to a NetCDF version change.

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 19 Apr 2011 20:56 UTC**
We do have tests for this among the Ferret benchmarks. The tests for doing things with attributes do a lot of it.  However, I had installed the ncdump from the library for our systems. 

So many combinations to try!

**Comment by karlmsmith** [comment]
**Comment by steven.c.hankin on 21 Apr 2011 22:03 UTC**
I believe we can work around this problem by tweaking the routine CD_OPEN_OUT, which is where the file specified in SAVE/APPEND/FILE=outname.nc gets opened.  (If so, we can cancel the request for help that we made to Unidata.)  The current code is the following:

```
* does it exist already ?
      INQUIRE ( FILE = cdfname, EXIST = file_exists )
      do_append = append .AND. file_exists

* open or create the file
      IF ( do_append ) THEN
         cdfid = NCOPN( cdfname, NCWRITE, cdfstat )
         IF ( cdfstat .NE. NCNOERR ) GOTO 5100
         CALL CD_SET_MODE( cdfid, pcd_reset_data, status )
         IF ( status .NE. merr_ok ) RETURN
      ELSE
              [...snip...]
      ENDIF
```

I think we can fix this as follows:
  1. Inside of the IF (do_append) block above, check to see if cdfname matches one of the already opened netCDF dataset names (check ds_des_names and ds_type).  If there is a match, then do not open the data set (again).  Instead return the cdfid value found in sf_lunit for the open dataset.  Also record the fact that the dataset was already open in some Boolean variable somewhere, cuz it is needed for step 2.
  2.  At the end of CDF_LIST, do not close the file, if the Boolean from step 1 indicates that it had been previously open.  Also make sure that the file is left in the correct netCDF mode.

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 26 Apr 2011 16:28 UTC**
Unidata says they have a fix for this incompatibility and will be releasing an update soon to address it. The newest HDF5 formats, which are used by NetCDF4, were incompatible with previous ones.

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 2 May 2011 19:39 UTC**
This appears to be fixed in FERRET v6.7 (nc413snapshot), dated 4/28/11.
